src/box/memtx_sort_data.c

```diff
@@ -0,0 +1,676 @@
+			entry.key.space_id = space_rv->id;
+			entry.key.index_id = 0; /* Just for clarity. */
+			mh_memtx_sort_data_entries_put(msd->entries,
+						       &entry, NULL, NULL);
+		}
+	}
+
+	return msd;
+}
+
+void
+memtx_sort_data_delete(struct memtx_sort_data *msd)
+{
+	assert(msd->fp == NULL);
+	mh_memtx_sort_data_entries_delete(msd->entries);
+	free(msd->dirname);
+	free(msd->fname);
+	free(msd);
+}
+
+int
+memtx_sort_data_create(struct memtx_sort_data *msd)
+{
+	/* Write the XLOG header. */
+	struct xlog log;
+	xlog_clear(&log);
+	if (xdir_create_xlog(&msd->dir, &log, &msd->vclock) != 0)
+		return -1;
+	int fd;
+	if (xlog_close_reuse_fd(&log, &fd, false) != 0) {
+		xlog_discard(&log);
+		return -1;
+	}
+
+	/*
+	 * Continue as FILE so writes are buffered. The fdopen call does not
+	 * truncate file on "w" but allows writes to arbitrary offsets in the
+	 * file.
+	 */
+	msd->fp = fdopen(fd, "wb");
+	if (msd->fp == NULL) {
+		xlog_discard(&log);
+		return -1;
+	}
+
+	/* The overall tuple count is to be updated later. */
+	fprintf(msd->fp, "Cardinality: ");
+	if ((msd->cardinality_offset = ftell(msd->fp)) == -1)
+		goto fail;
+	fprintf(msd->fp, "%020lld\n", 0LL);
+
+	/* Write all the sort data entries. */
+	fprintf(msd->fp, "Entries: %u\n", mh_size(msd->entries));
+	mh_int_t i;
+	mh_foreach(msd->entries, i) {
+		struct memtx_sort_data_entry *entry =
+			mh_memtx_sort_data_entries_node(msd->entries, i);
+		fprintf(msd->fp, "%u/%u: ", entry->key.space_id,
+			entry->key.index_id);
+		if ((entry->offset_offset = ftell(msd->fp)) == -1)
+			goto fail;
+		fprintf(msd->fp, "%016llx, ", 0LLU);
+		if ((entry->psize_offset = ftell(msd->fp)) == -1)
+			goto fail;
+		fprintf(msd->fp, "%016llx, ", 0LLU);
+		if ((entry->len_offset = ftell(msd->fp)) == -1)
+			goto fail;
+		fprintf(msd->fp, "%020llu\n", 0LLU);
+	}
+	fprintf(msd->fp, "\n");
+	return 0;
+
+fail:
+	memtx_sort_data_discard(msd);
+	return -1;
+}
+
+int
+memtx_sort_data_close(struct memtx_sort_data *msd)
+{
+	if (fseek(msd->fp, msd->cardinality_offset, SEEK_SET) != 0)
+		return -1;
+	fprintf(msd->fp, "%020lld\n", (long long)msd->cardinality);
+	fclose(msd->fp);
+	msd->fp = NULL;
+	return 0;
+}
+
+void
+memtx_sort_data_discard(struct memtx_sort_data *msd)
+{
+	if (msd->fp == NULL)
+		return; /* Discard prior to creation. */
+
+	fclose(msd->fp);
+	msd->fp = NULL;
+	xlog_remove_file(msd->fname, 0);
+}
+
+bool
+memtx_sort_data_begin(struct memtx_sort_data *msd,
+		      uint32_t space_id, uint32_t index_id)
+{
+	assert(msd->curr_entry == NULL);
+	struct memtx_sort_data_key key = {space_id, index_id};
+	mh_int_t i = mh_memtx_sort_data_entries_find(msd->entries, key, NULL);
+	if (i == mh_end(msd->entries))
+		return false; /* Not included index. */
+
+	struct memtx_sort_data_entry *entry =
+		mh_memtx_sort_data_entries_node(msd->entries, i);
+	assert(!entry->is_committed);
+	if (fseek(msd->fp, 0, SEEK_END) != 0) {
+		say_error("%s: space %u: index #%u seek failed, file"
```

Comment:
Let's handle errors appropriately and not ignore invalid sort data files. Instead, an error should be thrown.

Reply:
Done.